### ISSUE TITLE
Next got extreme slow and freeze browser when use back function of browser

### ISSUE THREAD

**Issue by mrtk0y** [vercel/next.js#16925]
# Bug report
Use <Link> as navigate using the back function of browser make it's extreme slow 

## Describe the bug
Using <Link> as navigator is fine but when using the back Function of browser the website is freeze and cant any except wait or open new tab
Video problem : https://imgur.com/a/YFfiUsa
Image Component using <Link> :  https://imgur.com/a/XwJubNK

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:
1. Go to ' https://wowedu.net/list-offline-course '
2. Click on 'Any Items'
3. Click Back function of browser 
4. Repeat for 2-3times 
5. The website got freeze

Code :  From the list-offline-course to course detail using Link 
![image](https://user-images.githubusercontent.com/37510824/92459438-e682d380-f1f0-11ea-8b19-ba3c84f86f9a.png)
Then click back of browser

## Expected behavior
Back is normal like when navigate to another href using Link 

## Screenshots
Video problem : https://imgur.com/a/YFfiUsa
Image Component using <Link> :  https://imgur.com/a/XwJubNK

## System information

- OS: macOS
- Browser tested on chrome - safari
- Version of Next.js: lastest 
![image](https://user-images.githubusercontent.com/37510824/92454810-22b33580-f1eb-11ea-8e8a-50978e42d9c6.png)

- Version of Node.js: v12.17.0

**Comment by mrtk0y** [comment]
Code :  From the list-offline-course to course detail using Link 
![image](https://user-images.githubusercontent.com/37510824/92459438-e682d380-f1f0-11ea-8b19-ba3c84f86f9a.png)

Then click back of browser

Im deployed it to server
1. enter  https://wowedu.net/list-offline-course
2. click one of items card 
![image](https://user-images.githubusercontent.com/37510824/92459696-3497d700-f1f1-11ea-8dca-7f23c779171e.png)

3. Click back function of browser 
4. Click the 2nd card, then back, 
5. Click the 3rd card, then back
6. Website got freeze

**Comment by 0xJMR** [comment]
Hey @mrtk0y,

I stumbled upon the same issue and did some investigation. In my case, this issue was caused by subscribing to the route-change event in a component using `Router.events.on("routeChangeStart", callback)` without surrounding `useEffect` hook. As a consequence, whenever the component has been remounted, the callback has been registered once again causing an unwanted increase in the emitter's callback collection. This combined with how the `popstate` event is handled by Next.js / React caused an increase in CPU usage which ultimately locked the web browser. Using the `useEffect` hook with proper unsubscribing of the event upon unmounting seems to have solved this issue.

I hope this somehow points you or others experiencing the same / a similar problem in the right direction.
